### ISSUE TITLE
undefined method `base_path' for {"supports_reload"=>false}:ResourceMasher::ResourceMash

### ISSUE THREAD

**Issue by JustinAiken** [kwhrtsk/chef-errbit-server#1]
On the first (and subsequent) runs of this, it install rbenv/unicorn, but then crashes, log below.
- Chef: 12.4.1
- Knife Solo
- No attributes, just running the recipe straight..

```
Recipe: errbit-server::app
  * service_factory[errbit] action create
    * service_factory_provider[errbit] action install

      ================================================================================
      Error executing action `install` on resource 'service_factory_provider[errbit]'
      ================================================================================

      NoMethodError
      -------------
      undefined method `base_path' for {"supports_reload"=>false}:ResourceMasher::ResourceMash

      Cookbook Trace:
      ---------------
      /root/chef-solo/cookbooks-2/resource_masher/libraries/resource_masher.rb:48:in `method_missing'
      /root/chef-solo/cookbooks-2/service_factory/libraries/mixin_provider_unix.rb:100:in `unix_path_prep'
      /root/chef-solo/cookbooks-2/service_factory/libraries/provider_upstart.rb:44:in `install_service'
      /root/chef-solo/cookbooks-2/service_factory/libraries/provider.rb:55:in `action_install'
      /root/chef-solo/cookbooks-2/run_action_now/libraries/run_action_now.rb:36:in `block in run_action_now'
      /root/chef-solo/cookbooks-2/run_action_now/libraries/run_action_now.rb:35:in `each'
      /root/chef-solo/cookbooks-2/run_action_now/libraries/run_action_now.rb:35:in `run_action_now'
      /root/chef-solo/cookbooks-2/service_factory/providers/default.rb:41:in `block in class_from_file'

      Resource Declaration:
      ---------------------
      # In /root/chef-solo/cookbooks-2/service_factory/providers/default.rb

       42:     service_factory_provider new_resource.name do
       43:       action :install
       44:       factory new_resource
       45:       provider sub_provider_name
       46:     end
       47:   ))

      Compiled Resource:
      ------------------
      # Declared in /root/chef-solo/cookbooks-2/service_factory/providers/default.rb:42:in `block in class_from_file'

      service_factory_provider("errbit") do
        provider Chef::Provider::ServiceFactoryProvider::Upstart
        action [:nothing]
        retries 0
        retry_delay 2
        default_guard_interpreter :default
        service_name "errbit"
        factory # Declared in /root/chef-solo/cookbooks-2/errbit-server/recipes/app.rb:71:in `from_file'

      service_factory("errbit") do
        action [:create]
        retries 0
        retry_delay 2
        default_guard_interpreter :default
        declared_type :service_factory
        cookbook_name :"errbit-server"
        recipe_name "app"
        service_desc "Errbit Unicorn"
        exec "/opt/errbit/unicorn.sh"
        env_variables {"PORT"=>"3000", "RAILS_ENV"=>"production", "UNICORN_PID"=>"/opt/errbit/shared/pids/unicorn.pid", "ERRBIT_HOST"=>"errbit.example.com", "ERRBIT_PROTOCOL"=>"http", "ERRBIT_ENFORCE_SSL"=>"false", "CONFIRM_RESOLVE_ERR"=>"true", "ERRBIT_CONFIRM_ERR_ACTIONS"=>"true", "ERRBIT_USER_HAS_USERNAME"=>"false", "ERRBIT_USE_GRAVATAR"=>"true", "ERRBIT_GRAVATAR_DEFAULT"=>"identicon", "ALLOW_COMMENTS_WITH_ISSUE_TRACKER"=>"true", "SERVE_STATIC_ASSETS"=>"true", "SECRET_KEY_BASE"=>"f258ed69266dc8ad0ca79363c3d2f945c388a9c5920fc9a1ae99a98fbb619f135001c6434849b625884a9405a60cd3d50fc3e3b07ecd38cbed7406a4fccdb59c", "ERRBIT_EMAIL_FROM"=>"errbit@example.com", "ERRBIT_EMAIL_AT_NOTICES"=>"[0]", "ERRBIT_PER_APP_EMAIL_AT_NOTICES"=>"false", "ERRBIT_NOTIFY_AT_NOTICES"=>"[1,10,100]", "ERRBIT_PER_APP_NOTIFY_AT_NOTICES"=>"false", "MONGO_URL"=>"mongodb://localhost", "GITHUB_URL"=>"https://github.com", "GITHUB_AUTHENTICATION"=>"true", "GITHUB_ACCESS_SCOPE"=>"[repo]", "DEVISE_MODULES"=>"[database_authenticatable,recoverable,rememberable,trackable,validatable,omniauthable]", "RBENV_VERSION"=>"2.2.2", "LC_ALL"=>"C"}
        run_user "errbit"
        run_group "errbit"
        pid_file "/opt/errbit/shared/pids/unicorn.pid"
        var_subpath "errbit"
      end

        declared_type :service_factory_provider
        cookbook_name :"errbit-server"
      end


    ================================================================================
    Error executing action `create` on resource 'service_factory[errbit]'
    ================================================================================

    NoMethodError
    -------------
    service_factory_provider[errbit] (/root/chef-solo/cookbooks-2/service_factory/providers/default.rb line 42) had an error: NoMethodError: undefined method `base_path' for {"supports_reload"=>false}:ResourceMasher::ResourceMash

    Cookbook Trace:
    ---------------
    /root/chef-solo/cookbooks-2/resource_masher/libraries/resource_masher.rb:48:in `method_missing'
    /root/chef-solo/cookbooks-2/service_factory/libraries/mixin_provider_unix.rb:100:in `unix_path_prep'
    /root/chef-solo/cookbooks-2/service_factory/libraries/provider_upstart.rb:44:in `install_service'
    /root/chef-solo/cookbooks-2/service_factory/libraries/provider.rb:55:in `action_install'
    /root/chef-solo/cookbooks-2/run_action_now/libraries/run_action_now.rb:36:in `block in run_action_now'
    /root/chef-solo/cookbooks-2/run_action_now/libraries/run_action_now.rb:35:in `each'
    /root/chef-solo/cookbooks-2/run_action_now/libraries/run_action_now.rb:35:in `run_action_now'
    /root/chef-solo/cookbooks-2/service_factory/providers/default.rb:41:in `block in class_from_file'

    Resource Declaration:
    ---------------------
    # In /root/chef-solo/cookbooks-2/errbit-server/recipes/app.rb

     71: service_factory "errbit" do
     72:   action        :create
     73:   service_desc  "Errbit Unicorn"
     74:   exec          "/opt/errbit/unicorn.sh"
     75:   env_variables _environment
     76:   run_user      _user
     77:   run_group     _group
     78:   pid_file      node["errbit"]["pid_file"]
     79:   create_pid    false
     80: end
     81:

    Compiled Resource:
    ------------------
    # Declared in /root/chef-solo/cookbooks-2/errbit-server/recipes/app.rb:71:in `from_file'

    service_factory("errbit") do
      action [:create]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      declared_type :service_factory
      cookbook_name :"errbit-server"
      recipe_name "app"
      service_desc "Errbit Unicorn"
      exec "/opt/errbit/unicorn.sh"
      env_variables {"PORT"=>"3000", "RAILS_ENV"=>"production", "UNICORN_PID"=>"/opt/errbit/shared/pids/unicorn.pid", "ERRBIT_HOST"=>"errbit.example.com", "ERRBIT_PROTOCOL"=>"http", "ERRBIT_ENFORCE_SSL"=>"false", "CONFIRM_RESOLVE_ERR"=>"true", "ERRBIT_CONFIRM_ERR_ACTIONS"=>"true", "ERRBIT_USER_HAS_USERNAME"=>"false", "ERRBIT_USE_GRAVATAR"=>"true", "ERRBIT_GRAVATAR_DEFAULT"=>"identicon", "ALLOW_COMMENTS_WITH_ISSUE_TRACKER"=>"true", "SERVE_STATIC_ASSETS"=>"true", "SECRET_KEY_BASE"=>"f258ed69266dc8ad0ca79363c3d2f945c388a9c5920fc9a1ae99a98fbb619f135001c6434849b625884a9405a60cd3d50fc3e3b07ecd38cbed7406a4fccdb59c", "ERRBIT_EMAIL_FROM"=>"errbit@example.com", "ERRBIT_EMAIL_AT_NOTICES"=>"[0]", "ERRBIT_PER_APP_EMAIL_AT_NOTICES"=>"false", "ERRBIT_NOTIFY_AT_NOTICES"=>"[1,10,100]", "ERRBIT_PER_APP_NOTIFY_AT_NOTICES"=>"false", "MONGO_URL"=>"mongodb://localhost", "GITHUB_URL"=>"https://github.com", "GITHUB_AUTHENTICATION"=>"true", "GITHUB_ACCESS_SCOPE"=>"[repo]", "DEVISE_MODULES"=>"[database_authenticatable,recoverable,rememberable,trackable,validatable,omniauthable]", "RBENV_VERSION"=>"2.2.2", "LC_ALL"=>"C"}
      run_user "errbit"
      run_group "errbit"
      pid_file "/opt/errbit/shared/pids/unicorn.pid"
      var_subpath "errbit"
    end
```

**Comment by kwhrtsk** [comment]
Hi.
This cookbook depends on many other community cookbooks.
[service_factory](https://github.com/org-binbab/cookbook-service_factory) is one of them.
And service_factory depends on [resource_masher](https://github.com/org-binbab/cookbook-resource_masher).
I think this error is compatibility problem of resource_masher between chef version 11 and 12.
I am able to reproduce this error with chef 12, but it is not possible with chef 11.
If you can use chef 11, it is a simple workaround.

Maybe, this cookbook will work with chef 12 if it does not use service_factory.
I hope to try (if it is possible) to rewrite this cookbook to use monit_wrapper instead of service_factory in a few days.

**Comment by JustinAiken** [comment]
> This cookbook depends on many other community cookbooks.
> service_factory is one of them.
> And service_factory depends on resource_masher.
> I think this error is compatibility problem of resource_masher between chef version 11 and 12.
> I am able to reproduce this error with chef 12, but it is not possible with chef 11.
> If you can use chef 11, it is a simple workaround.

Yeah, after I'd posted this issue, I'd attempted to make it work by trying the service_factory changes from https://github.com/org-binbab/cookbook-service_factory/pull/4, but had no luck...  Can't go back to Chef 11, would break too many other things..

> I hope to try (if it is possible) to rewrite this cookbook to use monit_wrapper instead of service_factory in a few days.

Ah cool, that would be awesome! We're using monit anyways...

If you don't get around to it, I may try to fork this and add an option to just drop a plain ol' init.d script...

**Comment by kwhrtsk** [comment]
@JustinAiken 
I released version 1.1.0, and this version use monit_wrapper instead of service_factory.
I've tested this version CentOS 6.7 and Ubuntu 14.10.
Can you try this version?
`sample/` dir will be useful to understand how it works.

**Comment by JustinAiken** [comment]
Awesome, will try monday first thing :)

**Comment by JustinAiken** [comment]
Sweet, gets much farther along now!

I opened #2 and #3 that fixed a few small errors I see, stuck at #4 now..
